src/device/socket/vsock.rs

```diff
@@ -101,24 +139,185 @@ impl<'a, H: Hal, T: Transport> VirtIOSocket<'a, H, T> {
+            src_port: connection_info.src_port.into(),
+            dst_port: connection_info.dst.port.into(),
+            op: VirtioVsockOp::Rw.into(),
+            len: (buffer.len() as u32).into(),
+            buf_alloc: connection_info.buf_alloc.into(),
+            fwd_cnt: connection_info.fwd_cnt.into(),
+            ..Default::default()
+        };
+        self.send_packet_to_tx_queue(&header, buffer)
+    }
+
+    /// Receives the buffer from the destination.
+    /// Returns the actual size of the message.
+    pub fn recv(&mut self, buffer: &mut [u8]) -> Result<usize> {
+        let connection_info = self.connection_info()?;
+        let header = VirtioVsockHdr {
+            src_cid: self.guest_cid.into(),
+            dst_cid: connection_info.dst.cid.into(),
+            src_port: connection_info.src_port.into(),
+            dst_port: connection_info.dst.port.into(),
+            op: VirtioVsockOp::Rw.into(),
+            buf_alloc: connection_info.buf_alloc.into(),
+            fwd_cnt: connection_info.fwd_cnt.into(),
+            ..Default::default()
+        };
+        // TODO: We shouldn't need this send to trigger the fill of the rx queue.
+        self.send_packet_to_tx_queue(&header, &[])?;
+        self.wait_one_in_rx_queue();
+        let packet = self.pop_packet_from_rx_queue()?;
+        match packet.hdr.op()? {
+            VirtioVsockOp::Rw => {
+                buffer
+                    .get_mut(0..packet.hdr.len())
+                    .ok_or(SocketError::BufferTooShort)?
+                    .copy_from_slice(packet.data);
+                Ok(packet.hdr.len())
+            }
+            VirtioVsockOp::Rst => Err(SocketError::ConnectionFailed.into()),
+            _ => Err(SocketError::InvalidOperation.into()),
+        }
+    }
+
+    /// Shuts down the connection.
+    /// TODO: Add a timeout for shutdown.
+    pub fn shutdown(&mut self) -> Result {
+        let connection_info = self.connection_info()?;
+        let header = VirtioVsockHdr {
+            src_cid: self.guest_cid.into(),
+            dst_cid: connection_info.dst.cid.into(),
+            src_port: connection_info.src_port.into(),
+            dst_port: connection_info.dst.port.into(),
+            op: VirtioVsockOp::Shutdown.into(),
+            ..Default::default()
+        };
+        self.send_packet_to_tx_queue(&header, &[])?;
+        let received_header = self.pop_header_only_packet_from_rx_queue()?;
+        match received_header.op()? {
+            VirtioVsockOp::Rst | VirtioVsockOp::Shutdown => {
+                info!("Disconnected from the peer");
+                self.connection_info = None;
+                Ok(())
+            }
+            _ => Err(SocketError::InvalidOperation.into()),
+        }
+    }
+
+    /// Waits until there is at least one element to pop in rx queue.
+    fn wait_one_in_rx_queue(&self) {
+        for _ in 0..10000000 {
+            if self.rx.can_pop() {
+                break;
+            }
```

Comment:
If you really need to spin waiting for this, at least call `core::hint::spin_loop()` inside the loop.